network/p2p/dns/cache.go

```diff
@@ -91,3 +97,39 @@ func (c *cache) updateTXTCache(txt string, addr []string) {
 		timestamp: runtimeNano(),
 	}
 }
+
+// invalidateIPCacheEntry atomically invalidates ip cache entry. Boolean variable determines whether invalidation
+// is successful.
```

Comment:
I'm still confused as to whether nodes should be expecting `NXDOMAIN`s or not as part of normal operations.

> one typical type of such errors is a no such a host error, i.e. when the DNS domain is no longer valid

That is a possible error but is it a realistic / common / typical scenario for a node? Under which normal circumstances would a node be trying to talk to a domain for x period of time, only to have the IP pulled from DNS while the node is still trying to actively talk to the domain?

Assuming that in normal operation there should never be an `NXDOMAIN` DNS response, why not just:
* Always return the cached IP when asked for the IP?
* If the domain has expired then do a single async lookup attempt.
* If `NXDOMAIN` then:
  * Don't delete anything and don't async lookup again until the IP is asked for again.
  * Log an error because we never expect `NXDOMAIN` when running?